doc/rtd/reference/datasources/nocloud.rst

```diff
@@ -4,97 +4,237 @@ NoCloud
 *******
 
 The data source ``NoCloud`` is a flexible datasource that can be used in
-multiple different ways. With NoCloud, one can provide configurations to
-the instance without running a network service (or even without having a
-network at all). Alternatively, one can use HTTP/HTTPS or FTP/FTPS to provide
-a configuration.
+multiple different ways.
 
-Configuration Methods:
+With NoCloud, one can provide configuration to the instance locally (without
+network access) or alternatively NoCloud can fetch the configuration from a
+remote server.
+
+Much of the following documentation describes how to tell cloud-init where
+to get its configuration.
+
+Runtime configurations
 ======================
 
-.. warning::
-    User data placed under ``/etc/cloud/`` will **not** be recognized as a
-    source of configuration data by the NoCloud datasource. While it may
-    be acted upon by cloud-init, using
-    :ref:`DataSourceNone<datasource_none_example>` should be preferred.
+Cloud-init discovers four types of configuration at runtime. The source of
+these configuration types is configurable with a meta-configuration. This
+meta-configuration can be delivered to cloud-init in different ways, but is
+different from the configurations that cloud-init uses to configure the
+instance at runtime.
+
+user-data
```

Comment:
For symmetry with vendor-data:

```suggestion
user-data (optional)
--------------------
```

Reply:
I agree that this symmetry would be nice, however it [that's not how it works](https://github.com/canonical/cloud-init/blob/550c685c98551f65c30832b186fe091721b48477/cloudinit/sources/DataSourceNoCloud.py#L92). This implementation choice was a little odd from a UI perspective, since user-data can actually be empty.

Maybe it would be best to standardize this interface and allow all of them to be optional?

Reply:
meta data and user data are sometimes optional depending on the source. It's maybe not worth documenting separately, but e.g., when using the kernel cmdline, I don't think you need either.

Reply:
> meta data and user data are sometimes optional depending on the source.

Thanks for the clarification @TheRealFalcon, and my apologies @aciba90 you were right too - we were describing the behavior of different sources. I guess I missed that they behave differently for different sources. That's.... not at all intuitive. It looks like wasn't always that way either - in https://github.com/canonical/cloud-init/commit/c1253945761c33bfa89289a63cdb8799fc18d019 user-data and meta-data changed from being optional to being required for labeled filesystems and for the local filesystem. Unfortunately the commit message doesn't even mention this behavior change so we have no justification for why these two sources behave differently.

> when using the kernel cmdline, I don't think you need either.

Confirmed. On further examination of the code it appears that all four types are optional for the following sources: kernel command line, DMI, system configuration at the top level[1], system configuration nested under the NoCloud key[2], HTTP, FTP.

Unless we think that there is a good reason to require I'm tempted to standardize this behavior so that we can have a more intuitive and user-friendly user-interface. Currently a local datasource that doesn't provide a meta-data file or a user-data file is already on a failure path, so just silently removing this superficial requirement would probably be safe to include in SRU. While this still would technically be required for users of nested system configuration in order for ds-identify to identify the system (in the case that no override is provided `ds=nocloud` isn't in the line configuration and `datasource_list` has >1 datasources included), at least cloud-init's Python code wouldn't be imposing a false requirement.

@aciba90 @TheRealFalcon Thoughts?

[1] note, however, that ds-identify nocloud detection is completely broken and always has been, however these configurations will still by applied by cloud-init if None is included in the datasource list (it usually is unless only a single datasource is specified)
[2] however ds-identify nocloud detection fails unless user-data and meta-data are both provided

Reply:
> I'm tempted to standardize this behavior so that we can have a more intuitive and user-friendly user-interface.

Just to confirm, you want to try to standardize on making user-data and meta-data optional for every possible ways of configuring NoCloud, right? If that's the case, I think it is worth the implementation and SRU effort.

Reply:
I've filed https://github.com/canonical/cloud-init/issues/5553 to track this.